File: frontend/src/icons.rs
//! Semantic icon vocabulary.
//!
//! Maps the app's UI concepts (`SAVE`, `DISPLAY`, `BASE`, …) to concrete
//! Material Symbols glyphs from `egui_material_icons`. Call sites refer to the
//! concept, never the raw glyph, so an icon choice is made once here and reused
//! consistently everywhere — and re-skinning a concept is a one-line change.

use eframe::egui;
pub(crate) use egui_material_icons::MaterialIcon;
use egui_material_icons::icons as mi;

/// Default color for icon glyphs: a dark gray, distinct from the fully-black body
/// text, so icons read as secondary chrome. Call sites that need a semantic tint
/// (e.g. destructive red) override it per icon.
pub(crate) const DEFAULT_COLOR: egui::Color32 = egui::Color32::from_gray(0x5A);

// Query sections.
/// The query's source table.
pub(crate) const BASE: MaterialIcon = mi::ICON_PSYCHIATRY;
/// An individual table, e.g. in the Base submenu's list of tables.
pub(crate) const TABLE: MaterialIcon = mi::ICON_TABLE;
pub(crate) const FILTER: MaterialIcon = mi::ICON_FILTER_ALT;
pub(crate) const SORT: MaterialIcon = mi::ICON_SWAP_VERT;
pub(crate) const DISPLAY: MaterialIcon = mi::ICON_KEY_VISUALIZER;
/// Full-querydown mode: the single-input raw-query editor and its toggle.
pub(crate) const QUERYDOWN: MaterialIcon = mi::ICON_CODE;
/// A query, shown in its tab handle and beside its entries in the explorer.
pub(crate) const QUERY: MaterialIcon = mi::ICON_MANAGE_SEARCH;
/// Pin/unpin a tab (VS Code "keep open"). Shown as the unpinned-tab indicator
/// button and in the pin/unpin menu items.
pub(crate) const PIN: MaterialIcon = mi::ICON_PUSH_PIN;

// Generic actions.
/// Overflow ("⋮") menu trigger.
pub(crate) const MORE: MaterialIcon = mi::ICON_MORE_VERT;
/// The query options ("build") menu trigger in the query builder toolbar —
/// configuring the query's base, presets, and other settings.
pub(crate) const BUILD: MaterialIcon = mi::ICON_BUILD;
/// The explorer/sidebar toggle when the explorer is open (closes it).
pub(crate) const EXPLORER_CLOSE: MaterialIcon = mi::ICON_LEFT_PANEL_CLOSE;
/// The explorer/sidebar toggle when the explorer is closed (opens it).
pub(crate) const EXPLORER_OPEN: MaterialIcon = mi::ICON_LEFT_PANEL_OPEN;
/// The superscript asterisk marking a query with unsaved changes.
pub(crate) const UNSAVED: MaterialIcon = mi::ICON_EMERGENCY;
pub(crate) const ADD: MaterialIcon = mi::ICON_ADD;
pub(crate) const RENAME: MaterialIcon = mi::ICON_EDIT;
pub(crate) const EDIT: MaterialIcon = mi::ICON_EDIT;
/// Duplicate a query into a new copy.
pub(crate) const DUPLICATE: MaterialIcon = mi::ICON_CONTENT_COPY;
pub(crate) const DELETE: MaterialIcon = mi::ICON_DELETE;
/// View the compiled SQL that a query would send to the query API.
pub(crate) const VIEW_SQL: MaterialIcon = mi::ICON_MANUFACTURING;
/// Remove an item, or close the now-playing bar.
pub(crate) const CLOSE: MaterialIcon = mi::ICON_CLOSE;
pub(crate) const CLEAR: MaterialIcon = mi::ICON_BACKSPACE;
pub(crate) const SAVE: MaterialIcon = mi::ICON_SAVE;
/// Revert an in-progress edit, or a saved query's unsaved changes back to its
/// last-saved version.
pub(crate) const REVERT: MaterialIcon = mi::ICON_UNDO;
/// (Re-)run the current query.
pub(crate) const RUN: MaterialIcon = mi::ICON_REFRESH;
/// Reload a list from the backend.
pub(crate) const REFRESH: MaterialIcon = mi::ICON_REFRESH;
/// Expanded disclosure arrow on a collapsible preset card.
pub(crate) const EXPAND_OPEN: MaterialIcon = mi::ICON_EXPAND_MORE;
/// Collapsed disclosure arrow on a collapsible preset card.
pub(crate) const EXPAND_CLOSED: MaterialIcon = mi::ICON_CHEVRON_RIGHT;

// Presets.
/// A hand-written ("custom") query fragment, shown on the "Custom …" rows of the
/// section options menus. (Material Symbols 0.6 has no `wand_shine`; this
/// magic-wand glyph is the closest match.)
pub(crate) const CUSTOM: MaterialIcon = mi::ICON_AUTO_FIX_HIGH;
/// A saved preset, shown beside preset entries.
pub(crate) const PRESET: MaterialIcon = mi::ICON_LINK;
/// The built-in "Shuffle" sorting preset, and its "Reshuffle" action.
pub(crate) const SHUFFLE: MaterialIcon = mi::ICON_SHUFFLE;
/// Manage the whole preset library ("toolbox").
pub(crate) const MANAGE_PRESETS: MaterialIcon = mi::ICON_HANDYMAN;

// Settings & command palette.
/// The Settings dropdown trigger at the bottom of the explorer.
pub(crate) const SETTINGS: MaterialIcon = mi::ICON_SETTINGS;
/// The Keyboard Shortcuts settings entry and its tab handle.
pub(crate) const KEYBOARD: MaterialIcon = mi::ICON_KEYBOARD_ALT;

// Playback.
pub(crate) const PLAY: MaterialIcon = mi::ICON_PLAY_ARROW;
pub(crate) const PAUSE: MaterialIcon = mi::ICON_PAUSE;
pub(crate) const NEXT: MaterialIcon = mi::ICON_SKIP_NEXT;
/// Scroll the results to the now-playing track.
pub(crate) const LOCATE: MaterialIcon = mi::ICON_MY_LOCATION;

/// The egui font family that renders Material Symbols glyphs. Derived from an
/// icon so it always matches the active style (filled vs outline) selected by
/// the crate's feature flags, keeping that choice to the one line in Cargo.toml.
pub(crate) fn family() -> egui::FontFamily {
    BASE.font_family()
}

/// A [`egui::FontId`] for painting an icon glyph at `size`.
pub(crate) fn font_id(size: f32) -> egui::FontId {
    egui::FontId::new(size, family())
}

File: frontend/src/icons.ts
// Semantic icon vocabulary.
//
// Mirrors `frontend-old-egui/src/icons.rs`: call sites name a UI *concept*
// (`Query`, `Close`, `Refresh`, …) and never a raw glyph, so an icon choice is
// made once here and reused everywhere. Each import is a build-time-inlined
// Solid SVG component (via unplugin-icons) that fills with `currentColor`, so
// the theme tokens style it and size comes from a `size-*` utility / font size.
//
// In Iconify's `material-symbols` set the base name is the *filled* variant
// (matching egui's filled Material Symbols); `-outline` would be the unfilled one.

import Query from "~icons/material-symbols/manage-search";
import ExplorerOpen from "~icons/material-symbols/left-panel-open";
import ExplorerClose from "~icons/material-symbols/left-panel-close";
import Close from "~icons/material-symbols/close";
import Add from "~icons/material-symbols/add";
import Refresh from "~icons/material-symbols/refresh";
import ExpandOpen from "~icons/material-symbols/expand-more";
import ExpandClosed from "~icons/material-symbols/chevron-right";
import Settings from "~icons/material-symbols/settings";
import Save from "~icons/material-symbols/save";
import Build from "~icons/material-symbols/build";
import Filter from "~icons/material-symbols/filter-alt";
import Sort from "~icons/material-symbols/swap-vert";
import Display from "~icons/material-symbols/key-visualizer";
import More from "~icons/material-symbols/more-vert";
import Custom from "~icons/material-symbols/auto-fix-high";
import Preset from "~icons/material-symbols/link";
import Unsaved from "~icons/material-symbols/emergency";
import Revert from "~icons/material-symbols/undo";
import Clear from "~icons/material-symbols/backspace";
import Shuffle from "~icons/material-symbols/shuffle";
import Delete from "~icons/material-symbols/delete";
import ViewSql from "~icons/material-symbols/manufacturing";
import Base from "~icons/material-symbols/psychiatry";
import Table from "~icons/material-symbols/table";
// One glyph, two concepts — as in `icons.rs`, where `RENAME` and `EDIT` are both
// `ICON_EDIT`.
import Rename from "~icons/material-symbols/edit";
import Edit from "~icons/material-symbols/edit";
import Duplicate from "~icons/material-symbols/content-copy";
import Play from "~icons/material-symbols/play-arrow";
import Pause from "~icons/material-symbols/pause";
import Next from "~icons/material-symbols/skip-next";
import Locate from "~icons/material-symbols/my-location";
import Keyboard from "~icons/material-symbols/keyboard-alt";

export const Icons = {
  Query, // query rows + tab handle icon
  ExplorerOpen, // sidebar toggle when closed
  ExplorerClose, // sidebar toggle when open
  Close, // tab close ×, opened-row ×
  Add, // new-tab (+)
  Refresh, // Queries-section reload + toolbar run
  ExpandOpen, // expanded section chevron
  ExpandClosed, // collapsed section chevron
  Settings, // static Settings footer
  Save, // toolbar save (shown while unsaved)
  Build, // wrench: query-actions menu trigger
  Filter, // Filter section toggle
  Sort, // Sort section toggle
  Display, // Display section toggle
  More, // ⋮ menu trigger (split button + custom input)
  Custom, // "Custom …" options-menu entry
  Preset, // preset tab / options-menu preset rows
  Unsaved, // red ✱ dirty-preset marker
  Revert, // undo: revert changes / revert preset edit
  Clear, // clear custom input
  Shuffle, // Shuffle built-in preset
  Delete, // delete (stubbed this session)
  ViewSql, // View SQL menu entry
  Base, // Change-base menu entry
  Table, // a table row in the base submenu
  Rename, // rename query (wrench menu + double-click)
  Edit, // edit a record (results context menu, record editor title)
  Duplicate, // duplicate query (wrench menu)
  Play, // now-playing bar: resume
  Pause, // now-playing bar: pause
  Next, // now-playing menu: skip to the next queued track
  Locate, // now-playing menu: jump to the playing track's row
  Keyboard, // shortcuts editor: "Record keys" search toggle
};
